geopolars/src/geodataframe.rs

```diff
@@ -16,4 +29,194 @@ impl GeoDataFrame for DataFrame {
         let geom_column = self.column("geometry")?;
         geom_column.convex_hull()
     }
+
+    fn spatial_join(
+        lhs: &DataFrame,
+        rhs: &DataFrame,
+        join_type: JoinType,
+        predicate: Predicate,
+        _lsuffix: Option<&str>,
+        _rsuffix: Option<&str>,
+    ) -> Result<DataFrame> {
```

Comment:
Ok so lots still left to do here (both implementing and tidying up the code) but this is an initial stab at the spatial join. The strategy is 

1. Construct the spatial indexes for the two DataFrames 
2. Use the indexes to get candidate pairs of geometries who's envelopes overlap 
3. For each pair actually perform the relevant geometry check, this depends on the types of both of the geometries and the indicated predicate. 
4. For each pair that we get a hit for, record their indexes (locations in the series)

This is where things get a little weird and I think we can figure out ways to improve what happens next.

5. We add a column to the left hand and right hand datasets which is just a series of the index of each entry 
6. We construct a DataFrame of the join index tuples we  constructed at the geometry checking phase.
7. We do the relevant join between the lhs DataFrame and the join DataFrame 
8. Do the relevant join between the rhs DataFrame and the join DataFrame 
9. Drop the temporary indexes

This works but it means generating a few extra columns temporarily. If we had access to polars "join with series"  (https://github.com/pola-rs/polars/blob/793fbf849400243f530696c9233eab791a1c0911/polars/polars-core/src/frame/hash_join/mod.rs#L663 ) methods, I think we could do it just with the tuples but those are private.

I dont think we want to implement out own join logic ( as the code for this in polars is pretty involved) So this kind of approach, where we generate the pairs of  things that need joined, and then use the existing join code is how we want to do it, but this implementation feels clunky.

Open to suggestions!

Reply:
@kylebarron would love your thoughts here !

Reply:
> 1. Construct the spatial indexes for the two DataFrames

Maybe this is a dumb question, but do we need to construct the spatial indexes for _both_ dataframes? Naively I would've thought that you could generate an index for just one. I.e. make a spatial index for the right dataframe and then iterate over every row in the left dataframe, checking against the tree of the right dataframe.

I don't have good ideas about the rest yet, but I agree that I'd love to not reimplement the low-level join code.

Reply:
I think for points and polygons, the point based spatial index isn't necessary.

For joins that involve two polygons datasets, I think there is some benefit to being able to quickly exclude non-overlapping envelopes. You could do this on iteration for sure, this is just a quick way of doing it. 

This was also just the quickest way to get a set of candidates in code  and test out the join proportion of the code. We can potentially optimize it for different cases as we go.

We can also make the code not generate the spatial indexes and instead just take as arguments. We could then use whatever indexes we have or manually iterate with one index, or manually iterate with no indexed depending on what's passed in.